### PR TITLE
refactor: extract leaf data child position logic

### PR DESCRIPTION
**Pull request by rphmeier**
this extracts the leaf data position logic, which we use to determine where to store the `LeafData`
corresponding to a particular leaf. this is in preparation for moving `seek` into a separate
`Seeker` struct.

### CODE REVIEW

**Comment by rphmeier** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/146?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#145** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/145?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#146** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/146?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#137** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/137?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: core/src/trie_pos.rs [hunk 1]
```diff
@@ -262,6 +271,26 @@ impl fmt::Debug for TriePosition {
     }
 }
 
+/// Child node indices.
```
Comment (pepyakin) [inline core/src/trie_pos.rs:274]: indices in which space?

File: nomt/src/page_cache.rs [hunk 2]
```diff
@@ -171,6 +179,32 @@ impl fmt::Debug for Page {
     }
 }
 
+/// Given a trie position and a current page corresponding to that trie position (None at the root)
+/// along with a function for synchronously loading a new page, get the page and indices where the
+/// leaf data for a leaf at `trie_pos` should be stored.
+pub fn leaf_data_positions(
```
Comment (pepyakin) [inline nomt/src/page_cache.rs:185]: Nit: feels misplaced. No better suggestion

File: nomt/src/page_cache.rs [hunk 3]
```diff
@@ -171,6 +179,32 @@ impl fmt::Debug for Page {
...
+    trie_pos: &TriePosition,
+    current_page: Option<&(PageId, Page)>,
+    load: impl Fn(PageId) -> Page,
+) -> (Page, PageId, ChildNodeIndices) {
+    match current_page {
+        None => {
```
Comment (pepyakin) [inline nomt/src/page_cache.rs:191]: `assert!(trie_pos.is_root())`?

File: nomt/src/cursor.rs [hunk 4]
```diff
@@ -280,64 +280,36 @@ impl PageCacheCursor {
...
         self.mode.with_read_pass(|read_pass| trie::LeafData {
-            key_path: page.node(&read_pass, left_idx),
-            value_hash: page.node(&read_pass, right_idx),
+            key_path: page.node(&read_pass, children.left()),
+            value_hash: page.node(&read_pass, children.right()),
         })
     }
 
     fn write_leaf_children(&mut self, leaf_data: Option<trie::LeafData>) {
+        let (page, page_id, children) = crate::page_cache::leaf_data_positions(
```
Comment (pepyakin) [inline nomt/src/cursor.rs:296]: seems like a misnomer tbh

**Review by pepyakin — APPROVED** [review]

**Comment by pepyakin** [comment]
### Merge activity

* **Apr 22, 10:08 AM EDT**: @pepyakin started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thrumdev/nomt/146).

File: nomt/src/page_cache.rs (hunk 2 above)
Comment (rphmeier) [reply nomt/src/page_cache.rs:185]: yes, I agree - also no better suggestions right now.

**Pull request merged**